src/json.js

```diff
@@ -135,56 +135,56 @@ export function tryParseJson(json, opt_onFailed) {
 }
 
 /**
- * Recursively checks strict equality of items in nested arrays and objects.
+ * Deeply checks strict equality of items in nested arrays and objects.
  *
  * @param {JSONValueDef} a
  * @param {JSONValueDef} b
- * @param {number} depth The maximum recursion depth. Must be finite.
+ * @param {number} depth The maximum depth. Must be finite.
  * @return {boolean}
  * @throws {Error} If depth argument is not finite.
  */
-export function recursiveEquals(a, b, depth = 3) {
-  if (!isFinite(depth)) {
-    throw new Error('depth arg must be finite: ' + depth);
+export function deepEquals(a, b, depth = 5) {
+  if (!isFinite(depth) || depth < 0) {
+    throw new Error('Invalid depth: ' + depth);
   }
   if (a === b) {
     return true;
   }
-  // Only check shallow equality for depth < 1.
-  if (depth < 1) {
-    return false;
-  }
-  if (typeof a !== typeof b) {
-    return false;
-  }
-  if (Array.isArray(a) && Array.isArray(b)) {
-    if (a.length !== b.length) {
-      return false;
-    }
-    for (let i = 0; i < a.length; i++) {
-      if (!recursiveEquals(a[i], b[i], depth - 1)) {
+  /** @type {!Array<{a: JSONValueDef, b: JSONValueDef, depth: number}>} */
+  const queue = [{a, b, depth}];
+  while (queue.length > 0) {
+    const {a, b, depth} = queue.shift();
+    // Only check deep equality if depth > 0.
+    if (depth > 0) {
+      if (typeof a !== typeof b) {
         return false;
+      } else if (Array.isArray(a) && Array.isArray(b)) {
+        if (a.length !== b.length) {
+          return false;
+        }
+        for (let i = 0; i < a.length; i++) {
+          queue.push({a: a[i], b: b[i], depth: depth - 1});
+        }
+        continue;
+      } else if (a && b && typeof a === 'object' && typeof b === 'object') {
```

Comment:
is the a && b check required? typeof a === 'object' && typeof b === 'object' should suffice. Otherwise similar check should be added in initial typeof check in beginning of while loop, adter the depth check.

Reply:
Yea, because: `typeof null === 'object'`.